### PR TITLE
Fix bad codegen in lambda+local function case

### LINKED ISSUE #29340: BadImageFormatException on calling generic local function
**Version Used**: 
current Visual Studio 2017 15.8.0
tested with:
- Console Application
- .NET Framework 4.7.1.
- Any CPU

**Steps to Reproduce**:
Compile an run following code:
```c#
using System.Linq;
class Program
{
	static void Main(string[] args)
	{
		var n = 0;
		do
		{
			void Foo<T>() => args.Select(a => (n, default(T)));
			Foo<float>();
		} while (false);
	}
}
```
**Expected Behavior**:
Compile and run without errors.  

**Actual Behavior**:
Program compiles without error but throws BadImageFormatException when run.

### PR DESCRIPTION
**Pull request by agocke**
In one case with lambdas capturing a type parameter without capturing a
variable of that type, we were incorrectly dropping a closure frame
because we didn't think it was necessary (there were no captured
variables). This would be true, except for delegate caching, which acts
like a captured variable. Since the delegate caching field must live in
a closure frame and the delegate caching field could have the type of
the captured type parameter, we have to include the type parameter as a
captured variable if the delegate type of a lambda contains a captured
type parameter.

This case only appears with an intervening closure, since otherwise a
lambda will always generate at least one closure class and the type
parameter would be captured anyway.

Fixes #29340

### CODE REVIEW

**Comment by 333fred** [comment]
@agocke there are a bunch of test failures.

File: Compilers.sln [hunk 1]
```diff
@@ -438,15 +438,13 @@ Global
 		{1EE8CAD3-55F9-4D91-96B2-084641DA9A6C} = {A41D1B99-F489-4C43-BBDF-96D61B19A6B9}
 		{9508F118-F62E-4C16-A6F4-7C3B56E166AD} = {E35DA3D1-16C0-4318-9187-6B664F12A870}
 		{F5CE416E-B906-41D2-80B9-0078E887A3F6} = {E35DA3D1-16C0-4318-9187-6B664F12A870}
-		{A41D1B99-F489-4C43-BBDF-96D61B19A6B9} = {32A48625-F0AD-419D-828B-A50BDABA38EA}
```
Comment (jcouv) [inline Compilers.sln]: >{A41D1B99-F489-4C43-BBDF-96D61B19A6B9} = {32A48625-F0AD-419D-828B-A50BDABA38EA} [](start = 2, length = 79)

Was changing the SLN file intentional?

File: src/Compilers/Test/Utilities/CSharp/Microsoft.CodeAnalysis.CSharp.Test.Utilities.csproj [hunk 2]
```diff
@@ -8,7 +8,7 @@
     <OutputType>Library</OutputType>
     <RootNamespace>Microsoft.CodeAnalysis.CSharp.Test.Utilities</RootNamespace>
     <DefaultLanguage>en-US</DefaultLanguage>
-    <TargetFramework>netstandard1.3</TargetFramework>
+    <TargetFrameworks>$(RoslynPortableTargetFrameworks)</TargetFrameworks>
```
Comment (jcouv) [inline src/Compilers/Test/Utilities/CSharp/Microsoft.CodeAnalysis.CSharp.Test.Utilities.csproj]: ><TargetFrameworks>$(RoslynPortableTargetFrameworks)</TargetFrameworks> [](start = 4, length = 70)

Those CSPROJ changes seem unrelated. Probably should be a separate PR.

File: src/Test/Utilities/Portable/CompilationVerifier.cs [hunk 3]
```diff
@@ -78,7 +81,37 @@ internal Metadata GetMetadata()
...
+                        var typeDef = metadataReader.GetTypeDefinition(typeDefHandle);
+                        if (metadataReader.GetString(typeDef.Name) == typeName)
+                        {
+                            var disassembler = new ICSharpCode.Decompiler.Disassembler.ReflectionDisassembler(output, default);
+                            disassembler.DisassembleType(peFile, typeDefHandle);
+                            found = true;
+                            break;
+                        }
+                    }
+                    Assert.True(found, "Could not find type named " + typeName);
```
Comment (jcouv) [inline src/Test/Utilities/Portable/CompilationVerifier.cs]: >Assert.True(found, "Could not find type named " + typeName); [](start = 20, length = 60)

nit: It would be great to print out the available types.

File: src/Test/Utilities/Portable/CompilationVerifier.cs [hunk 4]
```diff
@@ -78,7 +81,37 @@ internal Metadata GetMetadata()
...
+                }
+            }
+            AssertEx.AssertEqualToleratingWhitespaceDifferences(expected, output.ToString(), escapeQuotes: false);
```
Comment (jcouv) [inline src/Test/Utilities/Portable/CompilationVerifier.cs]: >escapeQuotes: false [](start = 93, length = 19)

nit: feels like `escapeQuotes` should be true

File: src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.Tree.cs [hunk 5]
```diff
@@ -324,11 +324,13 @@ private class ScopeTreeBuilder : BoundTreeWalkerWithStackGuardWithoutRecursionOn
                 private void Build()
                 {
                     // Set up the current method locals
-                    DeclareLocals(_currentScope, _topLevelMethod.Parameters);
+                    DeclareVariables(_currentScope, _topLevelMethod.Parameters);
+                    // Add type parameters
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.Tree.cs]: >Add type parameters [](start = 23, length = 19)

Would probably be good to expand this comment

File: Compilers.sln (hunk 1 above)
Comment (jaredpar) [reply Compilers.sln]: Looks like you're changing hte folder structure here. Don't think this should be a part of the change.

---
In reply to: [220336965](https://github.com/dotnet/roslyn/pull/30063#discussion_r220336965) [](ancestors = 220336965)

File: src/Test/Utilities/Portable/Roslyn.Test.Utilities.csproj [hunk 6]
```diff
@@ -6,8 +6,7 @@
     <PlatformTarget>AnyCPU</PlatformTarget>
     <OutputType>Library</OutputType>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
-    <TargetFrameworks>netstandard1.3;$(RoslynPortableTargetFrameworks);net46</TargetFrameworks>
-    <DefineConstants>$(DefineConstants);DNX</DefineConstants>
+    <TargetFrameworks>$(RoslynPortableTargetFrameworks)</TargetFrameworks>
```
Comment (jaredpar) [inline src/Test/Utilities/Portable/Roslyn.Test.Utilities.csproj]: I think this is the wrong approach. This DLL deliberately targets netstandard1.3 in order to avoid multi-targeting all the places you just added. The design of this DLL is to provide a netstandard1.3 API set with a net46 and netcoreapp2.1 implementation.

**Comment by jcouv** [comment]
Marked this PR as "for personal review" since it wasn't updated in a while.

**Comment by agocke** [comment]
@jcouv Thanks, I'll get back to this eventually.
